### ISSUE TITLE
What meta fields should we use and what data should we put in them?

### ISSUE THREAD

**Issue by collina** [m-lab/ndt-javascript-client#4]
client.os.name is set to CLIWebsockets, but according to spec, we still have:

* "client.browser.name"	Web browser name with a version number (for example "MSIE 7.0",  "Firefox/3.6.18", etc.)
* "client.kernel.version"	Operating system kernel version (for example "2.6.35-30-server")
* "client.version"	NDT client version (for example "3.6.4")

What do we want to save (e.g. User Agent)?

**Comment by nkinkade** [comment]
I think "CLIWebsockets" should definitely be changed, since that is surely only relevant to the Node.js test client, and in any case is not even a valid OS identifier. Can we just do something like:

* client.os.name = navigator.oscpu;
* client.browser.name = navigator.userAgent;
* client.kernel.version = # not sure this can even be known in a browser environment, perhaps we could put browser version instead?
* client.version = # manually set

**Comment by critzo** [comment]
I agree with nkinkade's suggestions and I'd like to add that we should set
- connection_spec.client_application = NDTjs 
or similar.

**Comment by pboothe** [comment]
I am worried about navigator.userAgent uniquely identifying people. Is there a way we can scrub it to be only minimally informative?  e.g. converting "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/42.0.2311.82 Safari/537.36" to "Chrome"?

**Comment by nkinkade** [comment]
I agree that the entire userAgent string should not be submitted.  Though it would probably be useful to have browser name and version (e.g., Firefox 37, Chrome 42, etc.).

**Comment by critzo** [comment]
Couldn't you just use a substring function to strip the userAgent string to remove anything greater than X characters? I'm also unclear how a user agent string can be used to uniquely identify people. Also, we're already collecting full user agent strings in some clients now.

**Comment by technosopher** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

I don't want to get in the way here, but I'm very much in agreement
with Nathan that we should sanitize out as much of the User Agent
String as we possibly can.  To provide a concrete example of how/why
the user agent string can be problematic, it so happens that my user
agent string (when I'm in a snarky mood) is:

"Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:20.0; The NSA is
Unconstitutional) Gecko/20100101 Firefox/20.0"

I know I'm a ridiculous edge case, but the point is that User Agent
Strings can definitely be uniquely identifying, unless we go to some
lengths to strip out anything but the browser (which we'd probably
have to do through string matching, to be thorough).

Jordan McCarthy
Open Technology Institute @ New America
Public Key: 0xC08D8042 | 4A61 3D39 4125 127D 65EA DDC2 BFBD A2E9 C08D 80
42

On 04/23/2015 03:16 PM, critzo wrote:
> Couldn't you just use a substring function to strip the userAgent
> string to remove anything greater than X characters? I'm also
> unclear how a user agent string can be used to uniquely identify
> people. Also, we're already collecting full user agent strings in
> some clients now.
> 
> — Reply to this email directly or view it on GitHub 
> <https://github.com/m-lab/ndt-javascript-client/issues/4#issuecomment-
95689634>.
>
-----BEGIN PGP SIGNATURE-----
Version: GnuPG v1.4.11 (GNU/Linux)

iQEcBAEBCgAGBQJVOUsEAAoJEL+9ounAjYBC9f4IAJwr9eIVLkjaN9noY05WcDoh
FN1665pyfapas9KhgQUM/FXc03kYbHa9op5C0tzm4SQv6GlFTsFtcCySuqxHgVsW
CyJ2VnBr+2AbjDkWAO5Rw9WxiTEr8duOnl+xcTs0Q3ESRgDHnFvVKCnyhMehH9n0
CqEmMOwhgtGOtBkevBtFWiOuNz4AqaQBydlNl+JH/AR8nMWHtp4zPG2nJqawx4Um
4ZzC33eFbAW8Rk+uiutmNSO+0lbVeMDGP2QPqnC+VCwqm647r4fstcEikeoyjUZE
ePM5tkNrNm+HKqmiddYZ2Bl++zUuaRQZLPqVTvqq8cdgw2VafN65JBIo3qJxK9g=
=mGsC
-----END PGP SIGNATURE-----

**Comment by critzo** [comment]
It looks to me that clients already running out there are stripping out portions of the user agent string already, likely to deal with this issue. SO, if that's the case, why don't we just repurpose the code that's doing that within the JS client?

**Comment by nkinkade** [comment]
I think this was the purpose of the 4Kb of code that Collin was referring to which parses the userAgent string.  userAgent strings seem to be really variable and unwieldy.  Unfortunately, I'm not sure that any sort of simple substr function could reliably give us what we want.  As to how a userAgent string might constitute PII, especially in combination with IP:

https://panopticlick.eff.org/
